Lewiner marchin cubes algorithm

almarklein opened the pull request:
## Background

In https://github.com/scikit-image/scikit-image/pull/469#issuecomment-17273874 I promised to contribute an additional marching cubes algorithm. This code has been hiding inside the visvis project, which is a bad place, since visvis is essentially a pure python lib. 

> I'd be happy to create a PR for [another MC algorithm] I hope to have some time for that soon.

... 3 years later, I finally made the time to do it :)

## Description

This adds an additional marching cubes algorithm, which is an implementation of the method proposed by Lewiner et. al. [1] The algorithm does not have the ambiguities that the classic marching cubes algorithm has and produces topologically correct results. Which is nice for visualization, and e.g. when you want to use the result for 3D printing. Further this implementation is faster than the current implementation. 

## Changed / added files

New:
* `_marching_cubes_lewiner.py`: Python code for the algorithm
* `_marching_cubes_lewiner_cy.pyx`: Cython code for the algorithm. 
* `_marching_cubes_lewiner_lut.py`: Defines auto-generated lookup tables used by the algorithm
* `mc_meta/LookUpTable.h`: original definition of LUTs
* `mc_meta/createluts.py`: code used to generate LUTs
* `mc_meta/MarchingCubes.cpp`: C++ code that this was ported from, for future reference
* `mc_meta/visual_test.py`: small script that you can run to visualize surfaces side-by-side with visvis

Modified:
* `__init__.py`, `setup.py`, `bento.info`: to include the new stuff
* `_marching_cubes.py`: a few small tweaks
* `tests/test_marching_cubes.py`: add tests

## Checklist

- [x] Copy code from visvis and adapt to skimage
- [x] Docstrings
- [x] Clean style (could probably need some work)
- [x] Optimizations? My Cython is rusty
- [x] Ok to work internally with float32?
- [x] Ok to produce vertices in float32? (Originally it was float32, because then it can be uploaded to the GPU without converting first.)
- [x] Unit tests
- [x] Few small improvements to classic algorithm (docs, level arg, and and removed unnecessary line)
- [x] Add `gradient_direction` arg to new function
- [x] add readme to `mc_meta` subdir to explain the contents (e.g. C++ is for reference only)

## References
[1] Efficient implementation of Marching Cubes' cases with
topological guarantees. Thomas Lewiner, Helio Lopes, Antonio
Wilson Vieira and Geovan Tavares. Journal of Graphics Tools
8(2): pp. 1-15 (december 2003)

cc @JDWarner

almarklein commented on skimage/measure/_marching_cubes.py:
```diff
@@ -130,9 +142,6 @@ def marching_cubes(volume, level, spacing=(1., 1., 1.),
 
-    # Calculate gradient of `volume`, then interpolate to vertices in `verts`
-    grad_x, grad_y, grad_z = np.gradient(volume)
-
```
@JDWarner please double-check, but it looks like this was a leftover. I can recommend `pyflakes` to detect cases like this :)

JDWarner replied:
@almarklein You're right - this got left behind in a refactor moving the _correct_mesh_orientation functionality into a separate function.

These days my tools would catch it...

almarklein replied:
> These days my tools would catch it...

Orthogonal discussion, but I was hinting at whether skimage should use pyflakes in CI. Or is this not done because most code is pyflakes-unreachable Cython anyway?

JDWarner replied:
We actually tried it but it wasn't satisfactory. The issue was only partially Cython. I believe the biggest problems were with the test suite, combined with the fact that PEP8 recommends grouping terms in mathematical equations naturally instead of blindly adding spaces. There are a lot of places where something like the first line below got flagged

``` python
out = img*x + 2*y - 3*x*z          # Readable, PEP8 correct, flagged by pyflakes
out = img * x + 2 * y - 3 * x * z  # Unreadable, PEP8 incorrect, passes pyflakes
```

despite being correct and preferred by PEP8, because nobody's implemented a flake8 checker that cares about order of operations.

almarklein replied:
Yeah, that flake rule is useless. But note that with flake8 you can turn rules on and off in a .cfg file (not with pyflakes).

JDWarner commented:
Thanks, @almarklein! Great to see this here - I was tempted multiple times to port it from `visvis`, but refrained. 

From an API perspective, would it be better to simply use combine these under the umbrella of the `marching_cubes` function rather than have two separate functions? The new `marching_cubes_lewiner` function has the ability to fall back with a kwarg and most of the arguments are compatible.

This is a non-ambiguous mesh and is superior in pretty much every way to the original. In my view this should become the default.

almarklein commented:
I have considered this, and I just realized that I could add `gradient_direction` too (this just determines the direction of the triangles, right?), which would make the arguments backwards compatible. However, the new function returns 2 extra values. Not sure what we could do about that.

emmanuelle commented:
Hi @almarklein , great to see you here :-).

I have some concerns about the maintainability of this new feature, especially the inclusion of C++ code in scikit-image. I know it's not stated explicitly stated in the contribution guidelines (and if we agree on this point, we should really write it in the guidelines), but we are also a pure Python/Cython package. The choice to stick with Python/Cython is a distinctive feature of scikit-image, if one compares for example with other image processing packages such as mahotas. 

Would there be any way to port the algorithm to Cython only?

JDWarner commented:
@emmanuelle If I'm reading the description correctly, the port is done - C++ and header files are there _for reference_ with only Python/Cython is executed. We do ship some pure C for phase unwrapping in `skimage.restoration`, but no C++ code at present.

I'll try to give this a proper review soon. Sorry for the delay, @almarklein!

JDWarner commented:
@almarklein I have no problem working internally with float32, especially in Cython. Unless there is major mathematical complexity resulting in large compounded error, the effect on mesh accuracy should be negligible.

almarklein commented:
Hi @emmanuelle, Josh is correct, the algorithm is pure Cython. It was ported from C++ and I thought it would be good to include that source code for later reference (e.g. if there is ever a bug, we can check whether it was introduced in porting). I added a task to add a readme to this "auxillary" directory to explain this better.

almarklein commented:
Gentle ping @JDWarner :)

JDWarner commented:
@almarklein I do apologize, there are a lot of other demands on my time right now.

First, we need the new Cython source added to the Bento info file which should allow builds to pass. After that I have to carve out a chunk of time to sit down with this...

almarklein commented:
@JDWarner No need to apologize, I was only pinging to keep this on the radar :)  no hurry.

I added to bento.info.

I also added an extra arg to the new algorithm, which makes it compatible with the arguments of the original algorithm. The remaining issue that would hold back making this the default are the output arguments. Because 1) they produce float32 data (to save a conversion when displaying in OpenGL), and 2) the new algorithm produces 4 output values instead of 2 (adds `normals` and `values`).

stefanv commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,195 @@
+    
+    This is an implementation of:
+        
+        Efficient implementation of Marching Cubes' cases with
```
Add a references section?

stefanv commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,195 @@
+        return fun(vertices, faces, normals, values)
+
+
+def _toArray(args):
```
Follow PEP8 function naming here and for _getMCLuts

stefanv commented on skimage/measure/_marching_cubes_lewiner_luts.py:
```diff
@@ -0,0 +1,529 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2012, Almar Klein
```
Please remove copyright statements and add contributor names to CONTRIBUTORS.txt

stefanv commented on skimage/measure/mc_meta/visual_test.py:
```diff
@@ -0,0 +1,72 @@
+""" 
```
I'm not sure where this file should live.  How do we make sure it is runnable?  We probably need to update travis-CI, install visvis, render to an output buffer, and compare to a pre-existing image or hash of that image.

almarklein replied:
Getting this working on CI sounds painful (especially with OpenGL in the mix). In principle, the quantitative tests should be sufficient. The purpose of this file is merely to do some quick comparisons of the algorithms and verify output visually during development. I added it to the repo because it might be useful for others (or future me's) when making modifications to the algorithms.

JDWarner replied:
This is too much for a quick docstring example, and too technical (and 3D) for the example gallery. It's a very useful thing for dev-to-dev communication though. We don't have much that falls into this category, so not a lot of precedent. However, as it isn't publicly exposed, I'm OK having it here if @stefanv doesn't veto.

stefanv commented on skimage/measure/tests/test_marching_cubes.py:
```diff
@@ -74,5 +97,58 @@ def test_correct_mesh_orientation():
+            for ix in range(vol.shape[2]):
+                z, y, x = float(iz)*a+b, float(iy)*a+b, float(ix)*a+b
+                vol[iz,iy,ix] = ( ( 
+                    (8*x)**2 + (8*y-2)**2 + (8*z)**2 + 16 - 1.85*1.85 ) * ( (8*x)**2 +
```
Where does this formula come from?  A reference / description would be helpful!

almarklein replied:
Sorry, it has been a while. I am pretty sure it's from Lewiner's code somewhere, but I cannot recall nor find the exact source.

stefanv commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,195 @@
+    
+    Parameters
+    ----------
+    volume : (M, N, P) array (the data is internally converted to
```
Remove bracketed description to paragraph text

stefanv replied:
Specify dtype

stefanv commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,195 @@
+        Whether to allow degenerate (i.e. zero-area) triangles in the
+        end-result. Default True. If False, degenerate triangles are
+        removed, at the cost of making the algorithm slower.
+    use_classic : bool
```
Do we keep the original implementation, or add a wrapper that uses this function with ``use_classic=True``?

almarklein replied:
I am not sure. The main issue seems to be the output arguments. If we can resolve that in an elegant way, I think we can replace it. Though I could not make this alg with `use_classic` produce the exact same results on all data.

stefanv commented on skimage/measure/tests/test_marching_cubes.py:
```diff
@@ -74,5 +97,58 @@ def test_correct_mesh_orientation():
+    assert _same_mesh(vertices1, faces1, vertices2, faces2)
+    assert _same_mesh(vertices1, faces1, vertices3, faces3)
+
+
```
Should we add another test for use_classic vs the old implementation of marching cubes?  The test case above seems to pass with or without use_classic.

almarklein replied:
In this test the data is simple enough for the algs to produce the same results. Below is a more complex use-case, but the old algorithm and new-with-use-classic do not produce the same result.

stefanv commented on skimage/measure/tests/test_marching_cubes.py:
```diff
@@ -74,5 +97,58 @@ def test_correct_mesh_orientation():
+    assert _same_mesh(vertices1, faces1, vertices3, faces3)
+
+
+def _same_mesh(vertices1, faces1, vertices2, faces2):
```
Why the integer rounding scheme, and not a floating point comparison with tolerance?

stefanv commented:
This is a sizable piece of work!  Thanks, @almarklein !

almarklein commented:
Thanks for the comments @stefanv. I addressed / responded to each of them.

I also looked into how the old alg and the classic new differ. They do produce significant different (i.e. not just slight difference in vertex position) results. This can also be seen from visually inspected the double-donut example; in the old alg the donuts touch and have holes at these locations, in the new-classic the donuts touch but are closed, and in the new alg the donuts do not touch. This probably mean that the new-classic is not *exactly* the classic MC. 

I don't have a strong opinion on whether the new alg should replace the old or not, but the fact that the algs produces slightly different results and have different output args make me lean slightly towards keeping both.

stefanv commented:
Thanks for looking into that, Almar.  I think a comment in the docstring highlighting this fact should be enough (that it imitates the classic mode, but isn't quite the same).

Does Lewiner's original code have a license?  Should we credit him as well?

almarklein commented:
@stefanv I dug up the email conversation that I had with the original author (Thomas Lewiner), and send him a notice. I cc-ed you so you can see what he said about the license (he considered it license-less and was ok with us using it as long as he was acknowledged). Incidentally, that also revealed the source of the double-torus formula (he emailed it to me).

almarklein commented:
I updated the docs, added appropriate acknowledgements, and fixed a Python 2 syntax error. But now we have a merge conflict. Is it policy to merge master in, or to rebase on master?

soupault commented:
@almarklein rebasing is the way to go.

almarklein commented:
I rebased, but appveyor seems to have been unable to finish, and where is Travis? Maybe because Travis is confused there were no real new commits?

jni commented:
@almarklein looks like a further rebase is needed on this! Would be awesome if you could push it through! =)

JDWarner commented on skimage/measure/_marching_cubes.py:
```diff
@@ -101,15 +109,19 @@ def marching_cubes(volume, level, spacing=(1., 1., 1.),
+    skimage.measure.marching_cubes_lewiner
     skimage.measure.mesh_surface_area
-
+    
```
When I wrote this, PEP 257 still called for an empty line at the end of docstrings. It no longer does, so let's just remove it.

JDWarner commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,204 @@
+from . import _marching_cubes_lewiner_cy
+
+
+
```
Only two newlines

JDWarner commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,204 @@
+    Lewiner marching cubes algorithm to find surfaces in 3d volumetric data
+    
+    In contrast to ``marching_cubes()``, this algorithm resolves
+    ambiguities and guarantees topologically correct results.
```
I'd actually be fine with an addition here stating something like "if you aren't sure which to use, use this one." It's that much better than the original algorithm.

JDWarner commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,204 @@
+        Step size in voxels. Default 1. Larger steps yield faster but
+        coarser results. The result will always be topologically correct
+        though.
+    allow_degenerate : bool
```
How inefficient is the pruning? It seems like most users would prefer having degenerate triangulations removed.

almarklein replied:
It takes between 10% and 50% of extra time. That's a lot of overhead, especially since the output might be the same, since degenerate triangles are rare in real world data,

JDWarner commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,204 @@
+    
+    Notes about the algorithm
+    -------------------------
+    
```
Section header should just be "Notes" to make the docscrape happy, and we can trim out this newline.

JDWarner commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,204 @@
+    --------
+    skimage.measure.marching_cubes
+    skimage.measure.mesh_surface_area
+
```
Trim this newline

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+            faces_ok_[j] = 0
+    
+    # Create mask and mapping to new vertex indices
+    vertices_ok = vertices_map1 == vertices_map0
```
Instead of keeping these both in memory and then doing a binary comparison at the end to see what changed, I think you could preallocate that binary array (`vertices_ok`) off the bat and set values to True or False as desired. That should increase efficiency a bit.

almarklein replied:
Mmm, I tried this today, but it fails, because it works recursively; the vertices_map1 must be kept
up to date during iteration.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+    
+    
+    cdef np.ndarray[INT32_T, ndim=1] vertices_map1_ = vertices_map1
+    cdef np.ndarray[INT32_T, ndim=1] faces_ok_ = faces_ok
```
For all the internal arrays defined above, you would likely gain significant speed using typed memoryviews instead. They are faster when directly indexing - slicing eliminates this advantage, but the below loop appears to be mostly directly indexed. For maximum efficiency the axis looping the fastest should be denoted for 2-D and higher arrays - for C-ordering, e.g., `double[:, ::1]`.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+    # we update the vertex map, i.e. we merge the corresponding vertices.
+    for j in range(faces_.shape[0]):
+        face_ = faces_[j]
+        i1, i2, i3 = face_[0], face_[1], face_[2]
```
Could these two lines be combined via 
``` python
i1, i2, i3 = faces_[j][0], faces_[j][1], faces_[j][2]
```
eliminating the `face_` variable entirely?

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+    for j in range(faces_.shape[0]):
+        face_ = faces_[j]
+        i1, i2, i3 = face_[0], face_[1], face_[2]
+        v1, v2, v3 = vertices_[face_[0]], vertices_[face_[1]], vertices_[face_[2]]
```
I think we can save a couple lookups here by immediately using `i1`, `i2`, and `i3` defined in the prior line.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+        i1, i2, i3 = face_[0], face_[1], face_[2]
+        v1, v2, v3 = vertices_[face_[0]], vertices_[face_[1]], vertices_[face_[2]]
+        if v1[0] == v2[0] and v1[1] == v2[1] and v1[2] == v2[2]:
+            vertices_map1_[i1] = vertices_map1_[i1] = imin(vertices_map1_[i1], vertices_map1_[i2])
```
Typo, if I'm reading this right: the second `vertices_map1_[]` should reference `i2`, not `i1`.

JDWarner replied:
This is the only type I can find, but may cause the difference in results for `use_classic`.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+        self.SUBCONFIG13 = Lut(SUBCONFIG13)
+
+
+def marching_cubes(im, double isovalue, LutProvider luts, int st=1, int classic=0):
```
We can type `im` and even make it a memoryview right here in the definition. We've already run `volume = np.ascontiguousarray(volume, np.float32)` back in the Python wrapper, so the type and the C-ordering is guaranteed in memory.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+    cdef int nt
+    cdef int case, config, subconfig
+    
+    # Unfortunately, using a step in the for loops degregades performance.
```
I wonder how much performance degradation was from recalculating `Nx-1` and `Ny-1` every loop? That's a lot of subtraction - unnecessary subtraction, as those can be precalculated. I'd hope Cython would translate `range(1, value, 1)` to the same code as `range(1, value)`. 

If I'm right, then these can probably be combined in a single loop.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+                    # Initialize cell
+                    cell.set_cube(isovalue, x, y, z, st,
+                        im_[z   ,y, x], im_[z   ,y, x+st], im_[z   ,y+st, x+st], im_[z   ,y+st, x],
+                        im_[z+st,y, x], im_[z+st,y, x+st], im_[z+st,y+st, x+st], im_[z+st,y+st, x] )
```
For even more speed, define internal variables for `z+st`, `y+st`, and `x+st` and calculate them in higher loops, only referencing them here. Calculate `x+st` only once in the innermost loop and reference it in all of these indexing operations.

JDWarner commented on skimage/measure/_marching_cubes_lewiner_cy.pyx:
```diff
@@ -0,0 +1,1463 @@
+                        # Determine number of vertices
+                        nt = 0
+                        while luts.CASESCLASSIC.get2(cell.index, 3*nt) != -1:
+                            nt += 1
```
Would this be equivalent? If so, it would save you a lot of multiplication in the innermost loop.
``` python
nt = 0
while luts.CASESCLASSIC.get2(cell.index, nt) != -1:
    nt += 3
```

JDWarner commented on skimage/measure/tests/test_marching_cubes.py:
```diff
@@ -74,5 +96,68 @@ def test_correct_mesh_orientation():
+    # New classic and new Lewiner are different
+    assert not _same_mesh(vertices2, faces2, vertices3, faces3)
+    # Would have been nice if old and new classic would have been the same
+    # assert _same_mesh(vertices1, faces1, vertices3, faces3, 5)
```
I found a single typo above in the Lewiner implementation's classic pathway - after fixing that, revisit this and see if they agree. That would certainly make life simpler...

JDWarner commented:
@almarklein sorry for the delay, I finally found the time to give this a good look. For the most part I only see minor style improvements, but there are some areas in the Cython with low-hanging-fruit optimizations possible - especially refactoring variables out of innermost loops and aggressive precalculating / referencing in the main `marching_cubes` Cython function. 

There are also some places where typed memoryviews could be used to good effect. Those gains are real, but secondary to refactoring and precalc/referencing. If memoryviews aren't your favorite thing, don't worry about them for now.

There does appear to be [a subtle typo in this line](https://github.com/scikit-image/scikit-image/pull/2052#discussion-diff-71091044R79) pertaining to the classic code path, which may explain the difference between the classic implementations.

almarklein commented:
@JDWarner Thanks for taking the time to review this! I was hoping to see you at SciPy, but then learned you weren't there. Nice to see it still got a good review *during* SciPy :) I will pick this up again ASAP.

almarklein commented:
All suggested doc tweaks are done.

> For all the internal arrays defined above, you would likely gain significant speed using typed memoryviews instead. 

Changed to using memoryviews. Certainly looks nicer. I don't really see a speedup though.

> Could these two lines be combined via `i1, i2, i3 = faces_[j][0], faces_[j][1], faces_[j][2]` eliminating the face_ variable entirely?

Yes.

> I think we can save a couple lookups here by immediately using i1, i2, and i3 defined in the prior line

Indeed. Why did I miss that? :/


> I wonder how much performance degradation was from recalculating Nx-1 and Ny-1 every loop? 
> I'd hope Cython would translate range(1, value, 1) to the same code as range(1, value).

Checked again, but nope, specifying a step in range() degrades performance by > 30%.
I managed to rewrite the loops using while loops, so we don't have to duplicate code
and still have fast code.

> For even more speed, define internal variables for z+st, y+st, and x+st and calculate them in the outermost possible loop level,

Did that. Does not seem to help performance. Compilers are unpredictable. Leaving your suggestion in-place though as it does not hurt the readability.

> Would this be equivalent? If so, it would save you a lot of multiplication in the innermost loop.
> `nt = 0; while luts.CASESCLASSIC.get2(cell.index, nt) != -1: nt += 3`

I don't think it would help. Its only a small loop, and we'd have to devide `nt` afterwards.

> Typo, if I'm reading this right: the second vertices_map1_[] should reference i2, not i1.

Good catch! Thanks!

Unfortunately this is not the (sole) cause for the difference between the two algorithms.

almarklein commented:
Travis is green. I Appveyor expected to fail? (it looks like a failure that I recently had in imageio, and was also present in mpl's CI, which was resolved by explicitly conda installing `msvc_runtime`.)

soupault commented:
@almarklein would you like to comment on this in the dedicated issue (https://github.com/scikit-image/scikit-image/issues/2207)? Thanks!

JDWarner commented:
This looks good to me. @scikit-image/core anyone else want to sign off?

I'm amazed Cython doesn't reduce `range(0, n, 1)` to the same code as `range(0, n)`! I wonder if that should be raised as an issue to them.

sciunto commented on skimage/measure/tests/test_marching_cubes.py:
```diff
@@ -74,5 +96,68 @@ def test_correct_mesh_orientation():
+    n = 48
+    a, b = 2.5/n, -1.25
+    isovalue = 0.0
+    #
```
Empty comment?

sciunto commented on skimage/measure/mc_meta/createluts.py:
```diff
@@ -0,0 +1,174 @@
+
+# Get base64 encode/decode functions
+PY3 = sys.version_info[0] == 3
+if PY3:
```
Merge those two lines?

jni replied:
Holy moly! Please read this and fix this bit:

http://astrofrog.github.io/blog/2016/01/12/stop-writing-python-4-incompatible-code/

Also, I would use `import six`; `if six.PY2:`. That'll make it easier to track down this code once we drop PY2 support.

almarklein replied:
Oh crap, I never write it this way anymore, but I suppose I did back when I wrote this :/

sciunto commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,203 @@
+    ..  [1] Thomas Lewiner, Helio Lopes, Antonio Wilson Vieira and Geovan
+            Tavares. Efficient implementation of Marching Cubes' cases with
+            topological guarantees. Journal of Graphics Tools 8(2)
+            pp. 1-15 (december 2003).
```
Please,add a DOI (`DOI:xxx`)

sciunto commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,203 @@
+                           gradient_direction='descent', step_size=1,
+                           allow_degenerate=True, use_classic=False):
+    """
+    Lewiner marching cubes algorithm to find surfaces in 3d volumetric data
```
Missing final point.

sciunto commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -0,0 +1,203 @@
+    In contrast to ``marching_cubes()``, this algorithm is faster,
+    resolves ambiguities, and guarantees topologically correct results.
+    Therefore, this algorithm generally a better choice, unless there
+    is a specific need for the classic algorithm.
```
Should we state that also by a note in the classic algorithm docstring?

almarklein replied:
Good point.

almarklein commented:
> I'm amazed Cython doesn't reduce range(0, n, 1) to the same code as range(0, n)! I wonder if that should be raised as an issue to them.

https://github.com/cython/cython/issues/532

almarklein commented:
Rethinking whether we should name this new algorithm `marching_cubes` and rename the previous variant to `marching_cubes_classic` (keeping it for reference, since it does provide slightly different results). The main thing holding this back is the different number of output arguments. Options:

* A - Mention change in release notes that users will need to change their code. People probably wont read those and have their code stop working. Probably a bad idea.
* B - Return the same data by default and add an input arg to return additional output arrays. People's code will keep working, unless its very bound to the exact results that are returned. Bit ugly long term solution though.
* C - Don't rename previous algorithm. New algorithm is named `marching_cubes_lewiner` as it is now, or maybe `marching_cubes2` which is much shorter to type, or perhaps `isosurface` which is equally correct (though inconsistent with the current name).

JDWarner commented:
The best algorithm should be put forward when several are available with a similar naming convention. If a user wants marching cubes, they should get our best marching cubes by running `marching_cubes` - not needing to dive into the literature or even docstrings to figure out `*_lewiner` is superior. We may not always be truly Pythonic with "one best way," because in science alternatives for comparison are important, but *we should make the best way the easiest way*.

Unfortunately, due to past choices this is not always possible without either a years-long deprecation or a breaking change. Some disagree with me on this. but I very strongly feel we should make the breaking change in such cases. Our API needs it.

So I'm in favor of option A, with a tweak. In addition to the release notes, we could add a deprecation warning to the name `marching_cubes` which tells users something like "A better,, more efficient algorithm is now used which now returns four instead of two outputs including normals (see docstring for details). Exact backwards compatibility with 0.12 and prior, is available with `marching_cubes_classic`." IMHO a breaking change with instruction is entirely different than slapping users with sudden `ValueError: too many values to unpack`

almarklein commented:
With a good warning like that, I think I can agree on breaking the API here. One problem though, is that after users adapt their code to use the renewed version correctly, the warning will continue to be displayed.

I just looked into using inspect to get the number of output arguments. Its obviously wrong to use that information for "logic", but I think it makes sense to use it to warn for API breakage unless we're certain that the user expects the new number of args. (the code works on Python 2.6 - 3.5, and can deal with `a, b* = ...`). How do you feel about that?

JDWarner commented:
I'm interested in how you'd do that with inspect - i tried to find some way to do that but output argument assignment seems to happen totally outside scope.

If we can check this, then absolutely the warning could only be displayed for the legacy pattern. That would definitely be the way to go.

almarklein commented:
Applied the rename. I mostly left `marching_cubes_lewiner` in place, but added `marching_cubes` that does the deprecation warning and then calls `marching_cubes_lewiner`. That leaves room for any future MC algs and makes testing a bit more explicit.

jni commented on skimage/measure/_marching_cubes_classic.py:
```diff
@@ -240,21 +242,21 @@ def correct_mesh_orientation(volume, verts, faces, spacing=(1., 1., 1.),
     warn(DeprecationWarning("`correct_mesh_orientation` is deprecated for "
-                            "removal as `marching_cubes` now guarantess "
-                            "correct mesh orientation."))
+                            "removal as `marching_cubes`_classic now "
```
Oops: this `_classic` should have been inside the single ticks. =)

jni commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -1,24 +1,55 @@
-def marching_cubes_lewiner(volume, level=None, spacing=(1., 1., 1.),
-                           gradient_direction='descent', step_size=1,
-                           allow_degenerate=True, use_classic=False):
+def expected_output_args():
```
Whoa! This is totally badass. =D It should probably have a leading underscore to indicate that it's a private function, though!

JDWarner replied:
Agreed, I'm impressed!

This may have other uses for similar deprecations in the future, so it might best fit in `skimage._shared`

jni commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -92,10 +123,32 @@ def marching_cubes_lewiner(volume, level=None, spacing=(1., 1., 1.),
+        nout = expected_output_args()
+    except Exception:
+        nout = 0
+    if nout <= 2:
```
I feel like this should be `== 2`: if someone wants to assign all the outputs to a tuple, they can do that correctly with either the old or the new signature, right?

almarklein replied:
Yes, if someone catches the result as a tuple, we have no way of knowing whether the change is going to break their code. Better safe than sorry, and warn in this case:
```python
mc_outputs = marching_cubes(...)
# ... somewhere later in the code
vertices, faces = mc_outputs
```

BTW: people could write compatible code (that does not raise the warning) by using `vertices, faces, *_ = marching_cubes(...)`.

jni commented on skimage/measure/_marching_cubes_lewiner.py:
```diff
@@ -92,10 +123,32 @@ def marching_cubes_lewiner(volume, level=None, spacing=(1., 1., 1.),
+    # This signature (output args) of this func changed after 0.12
+    try:
+        nout = expected_output_args()
+    except Exception:
```
This scares me: what kind of exceptions are we expecting here? Why can't this be more specific?

almarklein replied:
The `expected_output_args()` should be considered black magic. If it ever fails for whatever reason, I don't want it to break the MC algorithm ...

jni commented:
In case it wasn't clear from my comments, I totally dig the "safe option A" you've chosen. =) But there's a couple of tweaks to make to it.

almarklein commented:
Applied to of @jni 's suggestions. The other comments are now hidden because of the changes. I'll repeat them here:

> This scares me: what kind of exceptions are we expecting here? Why can't this be more specific?

The `expected_output_args()` function should be considered black magic. If it ever fails for whatever reason, I don't want it to break the MC algorithm ...

> I feel like this should be == 2: if someone wants to assign all the outputs to a tuple, they can do that correctly with either the old or the new signature, right?

Yes, if someone catches the result as a tuple, we have no way of knowing whether the change is going to break their code. Better safe than sorry, and warn in this case:
```python
mc_outputs = marching_cubes(...)
# ... somewhere later in the code
vertices, faces = mc_outputs
```

jni commented:
> The expected_output_args() function should be considered black magic. If it ever fails for whatever reason, I don't want it to break the MC algorithm ...

Got it. Maybe include this as a comment in the code for the next person? =)

> Yes, if someone catches the result as a tuple, we have no way of knowing whether the change is going to break their code. Better safe than sorry, and warn in this case:

Oh, I forgot that it was a warning and not an error. Fair enough!

JDWarner commented:
Does this look good to you now, @jni?

almarklein commented:
> Got it. Maybe include this as a comment in the code for the next person? =)

Done.

jni commented:
👍 !

JDWarner commented:
Awesome! Thanks for sticking out my delay in reviewing this, @almarklein, and for all your hard work on it.

almarklein commented:
Woohoo! Thanks all, for the great reviews.
